tools/loader/ImageClassifier.cpp

```diff
@@ -245,96 +346,90 @@ static void printTopKPairs(const std::vector<FloatIndexPair> &topKPairs) {
-  case ImageLayout::NCHW:
-    break;
-  case ImageLayout::NHWC:
-    Tensor dataNHWC;
-    data.transpose(&dataNHWC, NCHW2NHWC);
-    data = std::move(dataNHWC);
-    break;
-  }
+  GLOW_ASSERT(!(streamInputFilenamesMode && emittingBundle()) &&
+              "Cannot emit a bundle and also stream inputs.");
 
-  // The image name that the model expects must be passed on the command line.
-  const char *inputName = modelInputName.c_str();
+  // Use to make sure we only compile once.
+  bool isFirstRun = true;
 
-  // Create the model based on the input model format.
-  std::unique_ptr<ProtobufLoader> LD;
-  bool c2Model = !loader.getCaffe2NetDescFilename().empty();
-  if (c2Model) {
-    LD.reset(new Caffe2ModelLoader(
-        loader.getCaffe2NetDescFilename(), loader.getCaffe2NetWeightFilename(),
-        {inputName}, {&data.getType()}, *loader.getFunction()));
-  } else {
-    LD.reset(new ONNXModelLoader(loader.getOnnxModelFilename(), {inputName},
-                                 {&data.getType()}, *loader.getFunction()));
-  }
-
-  // Allocate tensors to back all inputs and outputs.
-  ctx.allocate(loader.getModule()->getPlaceholders());
-  if (convertInAndOutToFp16) {
-    // Convert the raw input to fp16.
-    data.convertToType(ElemKind::Float16Ty);
-    TypeAToTypeBFunctionConverter converter(
-        *loader.getFunction(), ElemKind::FloatTy, ElemKind::Float16Ty);
-    for (auto *placeholder : loader.getModule()->getPlaceholders()) {
-      converter.convertPlaceholder(*placeholder, &ctx);
+  // These will be set during the first run.
+  Placeholder *inputImagePH = nullptr;
+  Tensor *SMVarT = nullptr;
+  do {
+    // If streaming filenames, then load the next set of them.
+    if (streamInputFilenamesMode) {
+      getNextImageFilenames(&inputImageFilenames);
     }
-  }
-
-  // Get the Variable that the final expected Softmax writes into at the end of
-  // image inference.
-  Placeholder *SMVar = LD->getSingleOutput();
-  Tensor *SMVarT = ctx.get(SMVar);
 
-  // Create Variables for both possible input names for flexibility for the
-  // input model. The input data is mapped to both names. Whichever Variable is
-  // unused will be removed in compile().
-  auto *inputImage = llvm::cast<Placeholder>(LD->getNodeValueByName(inputName));
+    // Load and process the image data into the inputImageData Tensor.
+    Tensor inputImageData;
+    loadImagesAndPreprocess(inputImageFilenames, &inputImageData);
 
-  // Compile the model, and perform quantization/emit a bundle/dump debug info
-  // if requested from command line.
-  loader.compile(ctx);
+    // If this is the first run, then we need to build and compile the model.
+    if (isFirstRun) {
+      isFirstRun = false;
 
-  // If in bundle mode, do not run inference.
-  if (!emittingBundle()) {
+      // Build and compile the graph, and then get back the input Placeholder
+      // and output Softmax Tensor.
+      auto inputOutputPair = buildAndCompileAndGetInAndOutPair(
```

Comment:
Could you use an explicit type here instead of auto?
This is not obvious from the function name what are the types held by this pair.

Reply:
Done